keps/sig-cli/kubectl-socket-support.md

```diff
@@ -0,0 +1,174 @@
+over a unix socket. Currently we've only got `http` and `https`. `Docker` client does provide such supports via different types of Socket: `unix` and `tcp`.
+
+It would be great to be able to send requests from kubectl over sockets.
+
+### Goals
+
+* kubectl users can use sockets to connect to the apiserver:
+    * Allow passing socket file to `--server=unix:///path/to/socket` for kubectl, much like `-H unix:///var/run/docker.sock -H tcp://192.168.59.106` in `docker` client;
+    * Allow specifying sockets as server addresses in kubeconfig file;
+
+### Non-Goals
+
+* Convey credentials through unix sockets;
+* Adopting [SGX] to secure the memory access;
+
+[SGX]: https://www.intel.com/content/www/us/en/architecture-and-technology/software-guard-extensions.html
+
+## Proposal
+
+### User Stories
+
+To access a Kubernetes cluster, users will issue `kubectl` using an appropriate kubeconfig with scoped roles. However it is
+hard to avoid the kubeconfig files get spread, which will bring risks for production clusters. Moreover, it is not easy to revoke
+those kubeconfig files unless they get expired, especially TLS certificates.
+
+From the perspective of security, a more dynamic kubeconfig is preferred, such as using unix sockets, which could be revoked and closed
+as needed.
+
+Currently `kubectl` supports running proxy on a unix socket.
+
+```bash
+export TOKEN=USER1_TOKEN
+kubectl proxy --unix-socket=/home/user1/.kube/kube.sock --token=${TOKEN}
+```
+
+Issuing commands to any kube apiserver whereby unix socket is the preferred way to access it.
+
+```bash
+kubectl --server=/home/user1/.kube/kube.sock get pods
+```
+
+### Implementation Details/Notes/Constraints
+
+Also socket addresses can be specified as `Cluster.Server`.
+
+```go
+// staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go
+// Cluster contains information about how to communicate with a kubernetes cluster
+type Cluster struct {
+    // Server is the address of the kubernetes cluster (https://hostname:port).
+    // Socket address is still supported in experimental, such as unix:///var/run/kube.sock, tcp://192.168.59.106
+    Server string `json:"server"`
+    ...
+}
+```
+
+To connect to apiserver via sockets, we also need to modify current `TransportConfig` to convey unix sockets over HTTP.
+
+```go
+// staging/src/k8s.io/client-go/rest/transport.go
+// TransportConfig converts a client config to an appropriate transport config.
+func (c *Config) TransportConfig() (*transport.Config, error) {
+    conf := &transport.Config{
+    	...
+    	Dial: c.Dial,
+    }
+    
+    if scheme, sockpath, err := isSocketAddr(c.Host); err == nil {
+        conf.Dial = func(_ context.Context, _, _ string) (net.Conn, error) {
+            return net.DialTimeout(scheme, sockpath, defaultTimeout)
+        }
+    }   
+    ...
+}
+```
+
+### Risks and Mitigations
+
+There should be no compatibility issues. New clients will still be compatible with old servers and vice-versa.
```

Comment:
Discuss risks to other client-go config/transport consumers, especially ones used server-side that accept server addresses in config (admission webhooks, aggregated APIServices, conversion webhooks). At first glance, this would appear to allow existing API configurations to point those clients at socket files local to the master. Would that be allowed? If so, how would that increased exposure be communicated and mitigated by cluster deployers. If not, how would it be prevented?

Discuss the implications for other consumers of kubeconfig files (other languages, other clients, etc). Would socket support be added to other clients in https://github.com/kubernetes-client? Would support for sockets be incorporated into the client guidelines at https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/csi-new-client-library-procedure.md#development-process?

Reply:
> Discuss the implications for other consumers of kubeconfig files (other languages, other clients, etc). Would socket support be added to other clients in https://github.com/kubernetes-client? Would support for sockets be incorporated into the client guidelines at https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/csi-new-client-library-procedure.md#development-process?

This is still outstanding